tensorflow_estimator/python/estimator/experimental/tensor_forest.py

```diff
@@ -0,0 +1,266 @@
+        self._params,
+        tree_configs=tree_configs)
+
+  def inference_graph(self, dense_features):
+    """ Builds inference graphs"""
+    # Get predictions from all the trees
+    logits = [
+        gen_tensor_forest_ops.tensor_forest_tree_predict(
+            tree_variable,
+            dense_features,
+            self._params.logits_dimension)
+        for tree_variable in self._variables
+    ]
+
+    # shape of all_predict should be [batch_size, n_trees, logits_dimension]
+    all_predict = array_ops.stack(logits, axis=1)
+
+    average_values = math_ops.reduce_mean(
+        all_predict, name='logits', axis=1)
+    expected_squares = math_ops.reduce_mean(
+        all_predict * all_predict, axis=1)
+    logit_variance = math_ops.maximum(
+        0., expected_squares - average_values * average_values)
+
+    return average_values, logit_variance
+
+  def average_size(self):
+    sizes = [gen_tensor_forest_ops.tensor_forest_tree_size(tree_variable)
+             for tree_variable in self._variables]
+    return math_ops.reduce_mean(math_ops.to_float(array_ops.stack(sizes)))
+
+
+def _tensor_forest_model_fn(features,
+                            labels,
+                            mode,
+                            head,
+                            sorted_feature_columns,
+                            forest_hparams,
+                            config,
+                            name='tensor_forest'):
+  """Tensor forest model function"""
+  with ops.name_scope(name):
+    graph_builder = TensorForestGraphs(
+        forest_hparams, config)
+    # pylint: disable=protected-access
+    transformed_features = feature_column_lib._transform_features(
+        features, sorted_feature_columns)
+
+    dense_features = array_ops.concat(transformed_features.values(), axis=1)
+
+    logits, logits_variance = graph_builder.inference_graph(
+        dense_features)
+
+    summary.scalar('average_tree_size', graph_builder.average_size())
+
+  training_graph = None
+
+  def _train_op_fn(unused_loss):
+    del unused_loss
+    return training_graph
+
+  estimator_spec = head.create_estimator_spec(
+      features=features,
+      mode=mode,
+      labels=labels,
+      train_op_fn=_train_op_fn,
+      logits=logits)
+
+  logits_after_activation = estimator_spec.predictions[
+      prediction_keys.PredictionKeys.PREDICTIONS]
+
+  average_values = math_ops.reduce_mean(
+      logits_after_activation, name='avg_probs', axis=1)
+  expected_squares = math_ops.reduce_mean(
+      logits_after_activation * logits_after_activation, axis=1)
+  var_after_activation = math_ops.maximum(
+      0., expected_squares - average_values * average_values)
+
+  estimator_spec.predictions[VARIANCE_LOGITS_KEY] = logits_variance
+  estimator_spec.predictions[
+      VARIANCE_AFTER_ACTIVATION_KEY] = var_after_activation
+
+  return estimator_spec
+
+
+# @estimator_export('estimator.TensorForestClassifier')
+class TensorForestClassifier(estimator.Estimator):
```

Comment:
so there is a classifier but no regressor. I would suggest to either rename it to TensorForestEstimator, or adding a regressor as well (with l2 loss head)

Reply:
In boosted, we have a classifier, a regressor and an estimator that takes an arbitrary head...

Reply:
ah yeah, the regressor will be added. in a separate pr?

Reply:
sg!